### PR TITLE
EOS-17070 ISA-L: Integrate ISAL encode APIs and check performance.

### PR DESCRIPTION
**Pull request by sanjognaik**
EOS-17070 ISA-L: Integrate ISAL encode APIs and check performance.

- Added option 'run-ub' to m0 script to run unit benchmarking.
- Used Intel ISA arithmetic APIs in m0_parity_mul() and m0_parity_div() functions.
- Intel ISA APIs are accessed in user space only.
- Galois APIs are called only in kernel space for now.
- Integrated ISA encoding APIs
- Implemented isal_encode() and isal_recover() functions.
- Updated ub tests for parity_math.
- Increased number of iterations for parity_math ub tests to 100.
- Added function for differential parity calculation using Intel ISA API.

Signed-off-by: Sanjog Vikram Naik <sanjog.naik@seagate.com>

### CODE REVIEW

File: sns/parity_math.c [hunk 1]
```diff
@@ -670,6 +858,168 @@ static void xor_recover(struct m0_parity_math *math,
...
+	uint8_t *err_list = NULL;
+	uint8_t s;
+	int ret;
+
+	fail = (uint8_t*) fails->b_addr;
+	fail_count = fails_count(fail, unit_count);
+
+#if DEBUG
+	printf("\n fail_count: %d", fail_count);
+#endif
```
Comment (huanghua78) [inline sns/parity_math.c:887]: No need to use printf();

M0_LOG(M0_ALWAYS, ...) is good enough.
Or
m0_console_printf() is also OK.

File: sns/parity_math.c [hunk 2]
```diff
@@ -670,6 +858,168 @@ static void xor_recover(struct m0_parity_math *math,
...
+	}
+
+	for (ui = 1; ui < math->pmi_data_count; ++ui)
+		M0_ASSERT(block_size == data[ui].b_nob);
+
+	for (ui = 0; ui < math->pmi_parity_count; ++ui)
+		M0_ASSERT(block_size == parity[ui].b_nob);
+
+	M0_ALLOC_ARR(temp_matrix, (unit_count * math->pmi_data_count));
+	M0_ASSERT(temp_matrix != NULL);
```
Comment (huanghua78) [inline sns/parity_math.c:934]: All of these error (Out-of-memory) should be handled properly.
No M0_ASSERT() here.

Comment (sanjognaik) [reply sns/parity_math.c:930]: Created another ticket to return error codes for these functions.

File: sns/parity_math.c (hunk 1 above)
Comment (sanjognaik) [reply sns/parity_math.c]: Done

File: sns/parity_math.c [hunk 3]
```diff
@@ -297,12 +346,16 @@ M0_INTERNAL void m0_parity_math_fini(struct m0_parity_math *math)
...
+#endif /* __KERNEL__ */
 }
 
 M0_INTERNAL int m0_parity_math_init(struct m0_parity_math *math,
 				    uint32_t data_count, uint32_t parity_count)
 {
 	int ret;
+#ifndef __KERNEL__
+	uint32_t total_count;
+#endif /* __KERNEL__ */
```
Comment (huanghua78) [inline sns/parity_math.c:358]: We can define this variable in following section where it is used.

Usually we define a variable at the beginning of a function. But for simplicity, I think it is OK to do in the other way.

File: sns/parity_math.c [hunk 4]
```diff
@@ -500,6 +651,35 @@ static void reed_solomon_encode(struct m0_parity_math *math,
...
+{
+	uint32_t   pi; /* parity unit index. */
+	uint32_t   di; /* data unit index. */
+	uint32_t   block_size = data[0].b_nob;
+
+	math->data_frags[0] = (uint8_t *)data[0].b_addr;
+	for (di = 1; di < math->pmi_data_count; ++di) {
+		M0_ASSERT(block_size == data[di].b_nob);
+		math->data_frags[di] = (uint8_t *)data[di].b_addr;
+	}
```
Comment (huanghua78) [inline sns/parity_math.c:669]: Any reason to handle the "math->data_frags[0]" with exception?
Why not in the "for" loop and the "di" from 0?

File: sns/parity_math.c [hunk 5]
```diff
@@ -670,6 +854,168 @@ static void xor_recover(struct m0_parity_math *math,
...
+	m0_console_printf("\n err_list: ");
+#endif
+	for (i = 0, j = 0, r = 0; i < unit_count; i++) {
+		if (fail[i]) {
+			err_list[j] = i;
+
+			if (i < math->pmi_data_count)
+				data_out[j] = math->data_frags[i];
+			else
+				data_out[j] = math->parity_frags[i - math->pmi_data_count];
```
Comment (huanghua78) [inline sns/parity_math.c:907]: Here, the "math->data_frags[i]" and "math->parity_frags[xxx]" are not initialized to point to any buffer.


They are assigned in isa_encode(). But isa_encode and isa_recover() are independent from each other.

Comment (sanjognaik) [reply sns/parity_math.c]: Done.

File: sns/parity_math.c (hunk 4 above)
Comment (sanjognaik) [reply sns/parity_math.c:668]: In loop we are comparing block_size, which is data[0].b_nob, with b_nob of every data. If we start with di = 0, we are again comparing block_size with data[0].b_nob. Hence we started with di = 1.

File: sns/parity_math.c (hunk 3 above)
Comment (sanjognaik) [reply sns/parity_math.c]: Done.

**Comment by sanjognaik** [comment]
We have gathered performance data for encoding and decoding using Intel ISA APIs. This PR was just to get comments and get POC code reviewed.

**Pull request closed without merging**